File: ios/Classes/models/grid/FCPGridTemplate.swift
//
//  FCPGridTemplate.swift
//  flutter_carplay
//
//  Created by Oğuzhan Atalay on 21.08.2021.
//

import CarPlay

@available(iOS 14.0, *)
class FCPGridTemplate {
  private(set) var _super: CPGridTemplate?
  private(set) var elementId: String
  private var title: String
  private var tabTitle: String?
  private var systemIcon: String?
  private var showsTabBadge: Bool = false
  private var buttons: [CPGridButton]
  private var objcButtons: [FCPGridButton]

  init(obj: [String: Any]) {
    self.elementId = obj["_elementId"] as! String
    self.title = obj["title"] as! String
    self.tabTitle = obj["tabTitle"] as? String
    self.systemIcon = obj["systemIcon"] as? String
    self.showsTabBadge = obj["showsTabBadge"] as! Bool
    self.objcButtons = (obj["buttons"] as! [[String: Any]]).map {
      FCPGridButton(obj: $0)
    }
    self.buttons = self.objcButtons.map {
      $0.get
    }
  }

  var get: CPTemplate {
    let gridTemplate = CPGridTemplate.init(title: self.title, gridButtons: self.buttons)
    gridTemplate.elementId = self.elementId
    gridTemplate.tabTitle = tabTitle
    gridTemplate.showsTabBadge = showsTabBadge
    if let systemIcon = systemIcon {
      resolveTabIcon(systemIcon) { gridTemplate.tabImage = $0 }
    }

    self._super = gridTemplate
    return gridTemplate
  }

  public func update(with: any FCPTemplate) {
    guard let with = with as? FCPGridTemplate else {
      return
    }
  }
}

@available(iOS 14.0, *)
extension FCPGridTemplate: FCPTemplate {}

File: android/src/main/kotlin/com/oguzhnatly/flutter_carplay/models/grid/FCPGridTemplate.kt
package com.oguzhnatly.flutter_carplay.models.grid

import androidx.car.app.model.GridTemplate
import androidx.car.app.model.ItemList
import com.oguzhnatly.flutter_carplay.CPGridTemplate
import com.oguzhnatly.flutter_carplay.FCPRootTemplate

/**
 * A wrapper class for CPGridTemplate with additional functionality.
 *
 * @param obj A dictionary containing information about the grid template.
 */
class FCPGridTemplate(obj: Map<String, Any>) : FCPRootTemplate() {

    /// The underlying CPGridTemplate instance.
    private lateinit var _super: CPGridTemplate

    /// The title of the grid template.
    private var title: String

    /// An array of CPGridButton instances associated with the grid template.
    private var buttons: ItemList

    /// An array of FCPGridButton instances associated with the grid template.
    private var objcButtons: List<FCPGridButton>


    init {
        val elementIdValue = obj["_elementId"] as? String
        val titleValue = obj["title"] as? String
        val buttonsData = obj["buttons"] as? List<Map<String, Any>>
        assert(elementIdValue != null && titleValue != null && buttonsData != null) {
            "Missing required keys in dictionary for FCPGridTemplate initialization."
        }
        elementId = elementIdValue!!
        title = titleValue!!

        objcButtons = buttonsData!!.map {
            FCPGridButton(it)
        }

        val builder = ItemList.Builder()
        objcButtons.forEach {
            builder.addItem(it.getTemplate())
        }
        buttons = builder.build()
    }


    /** Returns the underlying CPGridTemplate instance configured with the specified properties. */
    override fun getTemplate(): CPGridTemplate {
        val gridTemplate = GridTemplate.Builder().setTitle(title).setSingleList(buttons)
        _super = gridTemplate.build()
        return _super
    }
}
